event-store-plans.md

```diff
@@ -0,0 +1,174 @@
+some nascent partition support in Weasel already. Need some easy ways for users to create their own partitioning strategies
+
+
+## IEventSession interface/service
+
+What if we had a new interface specifically for folks who really only use the event sourcing at different areas of the code that got you straight to 
+event sourcing behavior like:
+
+```csharp
+public interface IEventSession : IEventStore
+{
+    Task SaveChangesAsync(CancellationToken token = default);
+}
+```
+
+It'd be the `DocumentSession`, but with a more event store friendly interface so you don't have to do `session.Events.******` on everything. 
+
+
+## Document by document type identity map behavior
+
+Today, document sessions either have the identity map behavior for all document types or no identity map for all types. I think especially around the 
+projections, that it would be helpful to selectively turn on the identity map within the session for certain document types. I think this is actually pretty
+low hanging fruit because all you'd do is switch up the cached `IDocumentStorage` objects for a certain document type.
+
+
+## Ensure that aggregates updated "Inline" are coming from identity map
+
+Imagine this very common scenario:
+
+A CQRS command handler needs to:
+
+1. Fetch an aggregate of an event stream as a "write model" to process the command
+2. Emit some events
+3. Maybe update the projected aggregation if it's running `Inline`
+4. For whatever reason, access the current state of the aggregate again in the handler. Maybe because it's a response type in an HTTP handler, maybe to transmit
+   messages, who knows
+
+Maybe you even want the updated aggregate to make some kind of sense of the next part of the work -- which I'd at least push Wolverine folks to do with cascading messages, but still
+
+Regardless of the Marten `IDocumentSession` type, I think it would be valuable to **prevent** unnecessary extra reads of the aggregate, especially if there's some
+re-application of the events. 
+
+At a minimum, I think that the aggregation fetching should use the identity map *even if the session itself isn't using the identity map*. See the next item.
+This probably impacts both the `FetchForWriting()` behavior and the inline projection behavior
+
+This absolutely impacts the Wolverine "aggregate handler workflow". There's an opportunity to eliminate some serious inefficiency today
+
+
+## Versioned Projections for Blue/Green deployment
+
+This is really just sticking a "projection version number" on the Projection subclasses that would potentially be used as a suffix on the document
+tables for the projected documents (the document alias). This would enable blue/green deployment of the same projection so that nodes on the previous version can be 
+using the previous revision/definition of the projection while other nodes are running the newer model. 
```

Comment:
Would it still be based on the async daemon id? So deployment would need:
- new document class,
- new projection,
- new versioned API?

What if one would like to reuse the same document when:
- the structure didn't change, but there was a bug in apply method,
- Did they provide a new version of the document that's backwards compatible? Would they need to type document alias?

I think that we should give also an option to run it in the same node. So deploying a new version of the application that can handle both the old projection and the new one. That's needed for monoliths.

Reply:
As far as I can see, versioning only makes sense if:
- The document remains the same shape
- The changes to the apply methods would result different content in the document (eg adding an apply for an event that is already in use)
- OR: the devs have decided that a projection needs to be rebuilt for some undetectable reason (maybe they did a db restore or some old bug caused consistency problems)

If the document changes shape, then versioning would either require returning a different document type based on whether the new version has caught up (not a thing that C# supports) or migrating the old document shape to the new one which is worse because the resulting document is at best inaccurate at worst invalid until the rebuild is complete anyway.  It would be better to just have both deployed so that developers could land a change that switches their business logic to the new one once it is ready.

Finally, there are some situations where you would want to change the projection, but not rebuild/version the projection:
- Adding an apply for a new event that is not yet in your system.
- Removing an apply for an event that has been removed from your system (eg through restreaming, or versioning to another supported event)

Reply:
I agree that adding a new field with a sensible default value can be done with the same document, though only when that value doesnt need to be backfilled to be accurate... unless you are ok with it being inaccurate until the backing projection table is rebuilt... so many caveats